File: 0853-car-fleet/0853-car-fleet.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        int n = position.size();
        stack<int> mono_stack;
        map<int, float> arrival_time;
        
        for(int i=0;i<n;i++)    arrival_time[position[i]]=(target-position[i])/(1.0*speed[i]);
        
        sort(position.begin(), position.end());
        
        for(int pos: position){
            while(!mono_stack.empty() && arrival_time[pos]>=arrival_time[mono_stack.top()])
                mono_stack.pop();
            
            mono_stack.push(pos);
        }
        return mono_stack.size();
    }
};

File: 0853-car-fleet/0853-car-fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        n = len(position)
        arrival_time = {}
        for i in range(n):
            arrival_time[position[i]] = (target-position[i]) / speed[i]

        position.sort()
        mono= []

        for i in range(n):
            while mono and mono[-1] <= arrival_time[position[i]]:
                mono.pop() 
            mono.append(arrival_time[position[i]])
        return len(mono)